File: arrange.py
import os
import shutil
import subprocess

def get_folders_sorted():
    """Get a list of all folders in the current directory, sorted lexically."""
    current_dir = os.getcwd()
    folders = [f for f in os.listdir(current_dir) if os.path.isdir(os.path.join(current_dir, f))]
    folders.sort()
    return folders

def get_mp4_files_sorted(folders):
    """Get a list of .mp4 files from the given folders, sorted by directory and lexically."""
    mp4_files = []
    for folder in folders:
        folder_path = os.path.join(os.getcwd(), folder)
        files = [f for f in os.listdir(folder_path) if f.endswith('.mp4')]
        files.sort()
        mp4_files.extend([(folder, f) for f in files])
    return mp4_files

def copy_and_rename_files(mp4_files):
    """Copy .mp4 files to the current directory and rename them."""
    current_dir = os.getcwd()
    renamed_files = []
    for index, (folder, file_name) in enumerate(mp4_files, start=1):
        old_path = os.path.join(current_dir, folder, file_name)
        new_file_name = f"{index:02d}{file_name[file_name.find(' '):]}"  # Renaming logic
        new_path = os.path.join(current_dir, new_file_name)
        shutil.copy(old_path, new_path)
        renamed_files.append(new_file_name)
        print(f"Copied and renamed: {old_path} to {new_path}")
    return renamed_files

def create_ffmpeg_concat_file(renamed_files):
    """Create a temporary file for ffmpeg concatenation."""
    with open('ffmpeg_concat_list.txt', 'w') as f:
        for file_name in renamed_files:
            f.write(f"file '{file_name}'\n")

def main():
    folders = get_folders_sorted()
    write_to_folders_txt(folders)
    mp4_files = get_mp4_files_sorted(folders)
    renamed_files = copy_and_rename_files(mp4_files)
    create_ffmpeg_concat_file(renamed_files)

def write_to_folders_txt(folders):
    with open('folders.txt', 'w') as f:
        for folder in folders:
            f.write(f"{folder}\n")

if __name__ == "__main__":
    main()

File: src/bin/arrange.rs
use std::env;
use std::fs;
use std::io::Write;

fn get_folders_sorted() -> Vec<String> {
    let current_dir = env::current_dir().unwrap();
    let mut folders: Vec<String> = fs::read_dir(current_dir)
        .unwrap()
        .filter_map(|entry| {
            let entry = entry.unwrap();
            let path = entry.path();
            if path.is_dir() {
                Some(path.file_name().unwrap().to_str().unwrap().to_string())
            } else {
                None
            }
        })
        .collect();
    folders.sort();
    folders
}

fn get_mp4_files_sorted(folders: Vec<String>) -> Vec<(String, String)> {
    let mut mp4_files = Vec::new();
    for folder in folders {
        let folder_path = env::current_dir().unwrap().join(&folder);
        let mut files: Vec<String> = fs::read_dir(&folder_path)
            .unwrap()
            .filter_map(|entry| {
                let entry = entry.unwrap();
                let file_name = entry.file_name().into_string().unwrap();
                if file_name.ends_with(".mp4") {
                    Some(file_name)
                } else {
                    None
                }
            })
            .collect();

        for file in &mut files {
            if !file.chars().nth(0).unwrap().is_digit(10)
                || !file.chars().nth(1).unwrap().is_digit(10)
            {
                let new_file = format!("0{}", file);
                let old_path = folder_path.join(&file);
                let new_path = folder_path.join(&new_file);
                fs::rename(&old_path, &new_path).unwrap();
                *file = new_file;
            }
        }
        files.sort();
        for file in files {
            mp4_files.push((folder.clone(), file));
        }
    }
    mp4_files
}

fn copy_and_rename_files(mp4_files: Vec<(String, String)>) -> Vec<String> {
    let current_dir = env::current_dir().unwrap();
    let mut renamed_files = Vec::new();
    for (index, (folder, file_name)) in mp4_files.into_iter().enumerate() {
        let old_path = current_dir.join(&folder).join(&file_name);
        let new_file_name = format!(
            "{:03}{}",
            index + 1,
            &file_name[file_name.find(' ').unwrap_or(0)..]
        );
        let new_path = current_dir.join(&new_file_name);
        fs::rename(&old_path, &new_path).unwrap();
        renamed_files.push(new_file_name.clone());
        println!("Copied and renamed: {:?} to {:?}", old_path, new_path);
    }
    renamed_files
}

fn create_ffmpeg_concat_file(renamed_files: Vec<String>) {
    let mut file = fs::File::create("ffmpeg_concat_list.txt").unwrap();
    for file_name in renamed_files {
        writeln!(file, "file '{}'", file_name).unwrap();
    }
}

fn write_to_folders_txt(folders: Vec<String>) {
    let mut file = fs::File::create("folders.txt").unwrap();
    for folder in folders {
        writeln!(file, "{}", folder).unwrap();
    }
}

fn main() {
    let folders = get_folders_sorted();
    write_to_folders_txt(folders.clone());
    let mp4_files = get_mp4_files_sorted(folders);
    let renamed_files = copy_and_rename_files(mp4_files);
    create_ffmpeg_concat_file(renamed_files);
}
